### PR TITLE
Implement user handles

### LINKED ISSUE #75: add custom profile URLs
Allow users to have custom profile URLs, e.g. `https://members.acmucsd.com/user/sumeetbansal`. Will probably require an additional `customUrl` column in the `Users` table and modifying the existing `GET /api/v1/users/:uuid` route.

### PR DESCRIPTION
**Pull request by shanekim28**
Closes #75

### CODE REVIEW

**Comment by farisashai** [comment]
Completed functionality for user handles.

Necessary behaviors:
- Get user handle as a field in public profile to view your handle before editing it.
- Get user profile by handle.
- Patch handle field.
- Initialize default handle for all existing accounts and when new account is registered.

API Functionality
**Register Route Provides Valid Default Handle**
Screenshot 2023-02-16 at 9 43 33 PM

**Get My User Profile with Handle Field**
Screenshot 2023-02-16 at 9 44 02 PM

**Get User Profile by Handle**
Screenshot 2023-02-16 at 10 00 45 PM

**Patch User Handle**
Screenshot 2023-02-16 at 10 00 28 PM

**Review by farisashai — APPROVED** [review]
LGTM 🚀

File: tests/handles.test.ts
Comment (shravanhariharan2) [inline tests/handles.test.ts:31]: nit: for consistency with other tests, let's remove the comments here since all tests follow this structure, and let's make the test descriptions start with uppercase and every other word be lowercase instead of all uppercase

Comment (shravanhariharan2) [reply tests/handles.test.ts:31]: also nit: 'Logged In User Can Successfully Set Their Own Handle' -> 'Logged-in user can *update* their own handle' so that it's clear that what's being tested is the update rather than the "set" which can happen in registerUser

File: tests/handles.test.ts
Comment (shravanhariharan2) [inline tests/handles.test.ts:41]: Right, the decorators only run when the API is hit via. HTTP (my bad, I should probably add this to the testing guidelines). But what we can do is manually invoke the validator that we want to test, as done here: https://github.com/acmucsd/membership-portal/blob/master/tests/feedback.test.ts#L47-L56

File: tests/handles.test.ts
Comment (shravanhariharan2) [inline tests/handles.test.ts:60]: I'm a bit unsure what this means, but it seems like this and the next test are essentially testing the same thing, so I don't think we need this test unless there's something I'm missing

File: tests/handles.test.ts
Comment (shravanhariharan2) [inline tests/handles.test.ts:108]: nit: we don't need a newline between each line of code, we can just group them by their purpose (e.g. group 99 and 101 for making the controller call, and 103-107 for the assertions.

File: tests/handles.test.ts
Comment (shravanhariharan2) [inline tests/handles.test.ts:110]: What is this testing that hasn't been tested already? If this is testing finding a user's profile by their handle that isn't the current user, then I think only 1 API call / assertion is necessary instead of looping over multiple users/API calls, since those are essentially testing the same thing.

File: tests/handles.test.ts
Comment (shravanhariharan2) [inline tests/handles.test.ts:4]: nit: let's rename the file to `user.handles.test.ts` again to make the context clear of what "handles" refer to.

**Review by shravanhariharan2 — COMMENTED** [review]
Functionality looks good, most of my comments are nitpicks around formatting, conventions, and testing. Let me know if there are any questions with anything

File: api/validators/GenericRequests.ts
Comment (farisashai) [reply api/validators/GenericRequests.ts:22]: So tbh I was initially planning to set the default value for user handles to `{firstName}-{lastName}-{hash}` as we discussed, but then I realized that I couldn't use the existing validators for `IsAlphanumeric` if I included a dash and would have to create a custom `IsValidHandle` validator. 

I did try the approach of making a custom validator but couldn't figure out how to return different default error messages depending on whether the length constraint or regex was broken. If you know how to handle an error message for invalid length and invalid characters in the same validator and think it's worth changing, I can switch the default to include the dash. 

I do think that it would improve readability of handles as a bonus.

**Comment by farisashai** [comment]
How do we deal with upper/lower case handle values? Should they be treated as identical if the handle is "Faris-Ashai-123" and the API query param is "faris-ashai-123"?

**Comment by farisashai** [comment]
Note: I want to test that the default value is exactly 32 characters for a name that exceeds 32 char length also

File: tests/auth.test.ts
Comment (farisashai) [reply tests/auth.test.ts:137]: I followed the same syntax as the first test in this file `user can register a new account` where we use the admin account to query for the `getUser` to validate that the registration was successful and the profile can be viewed by an existing user. The account arguably doesn't need to be an `admin` in either of these test cases since we are fetching the public profile type. 

I just changed it to a be a standard user for the query with no functional difference - we may want to adjust the other test case also if this is a concern in any way.

File: api/validators/GenericRequests.ts
Comment (shravanhariharan2) [reply api/validators/GenericRequests.ts:22]: We can split both these validations into two separate validators - the `Length(3,32)` decorator, and a custom `HasValidSymbols` (or some similarly named validator) decorator that checks the dashes/regex.

File: api/decorators/Validators.ts
Comment (shravanhariharan2) [inline api/decorators/Validators.ts:52]: nit: let's add a small comment here on what characters get matched with this regex (just so that developers who might not know how to parse regex still understand what's happening here).

File: api/decorators/Validators.ts
Comment (shravanhariharan2) [inline api/decorators/Validators.ts:60]: Should we specify what are valid characters? Either here or on the frontend?

File: migrations/0035-add-handle-column-to-users-table.ts
Comment (shravanhariharan2) [inline migrations/0035-add-handle-column-to-users-table.ts:27]: nit: let's update this comment to add the dashes

Comment (shravanhariharan2) [reply migrations/0035-add-handle-column-to-users-table.ts:27]: Also, what if this default value is longer than 32 characters? Should we apply our length constraints here as well (similar to how we do in `UserAccountService`?)

File: services/UserAccountService.ts
Comment (shravanhariharan2) [inline services/UserAccountService.ts:42]:
```suggestion
      .findByHandle(handle));
    if (!user) throw new NotFoundError('No user associated with this handle was found');
    return user;
```

File: services/UserAccountService.ts
Comment (shravanhariharan2) [inline services/UserAccountService.ts:50]: nit: let's update the comment to add the dashes here as well. Also, I think the param and return doc comments aren't really necessary, and since we don't include them anywhere else, let's remove them.

File: services/UserAuthService.ts
Comment (shravanhariharan2) [inline services/UserAuthService.ts:37]: Do we need the optional chaining here? I'm not sure there's ever a case where `registration` is null or undefined. If we don't need it, let's remove it

File: tests/auth.test.ts
Comment (shravanhariharan2) [inline tests/auth.test.ts:118]: nit: I think we should implement a `UserRegistrationFactory` to generate these fields for us, so that the tests are shorter and make it more clear which fields we are intentionally setting vs. which ones are arbitrary. E.g., every field in this object doesn't matter except handle, so we should have something instead like:
```ts
const user = UserRegistrationFactory.fake({ handle: existingMember.handle });
```

File: tests/auth.test.ts
Comment (shravanhariharan2) [inline tests/auth.test.ts:126]: nit: it's a bit unclear at a glance what the purpose of `FactoryUtils.randomHexString()` is here. Let's initialize a properly named variable for it instead and pass that variable into here.

File: tests/auth.test.ts
Comment (shravanhariharan2) [inline tests/auth.test.ts:132]:
```suggestion

  test('User registration with a full name longer than 32 characters truncates handle to exactly 32 characters', async () => {
```

File: tests/auth.test.ts
Comment (shravanhariharan2) [inline tests/auth.test.ts:147]: nit: instead of hardcoding a long firstName and lastName here, let's instead generate a random string of a specified length so that it's clearer what length of the full name is. e.g.:
```ts
const user = UserRegistrationFactory.fake({
  firstName: faker.datatype.string(20),
  lastName: faker.datatype.string(20),
})
```

File: tests/auth.test.ts
Comment (shravanhariharan2) [inline tests/auth.test.ts:199]: Since we only care about the handle in this test, we only need to assert that the handle is what we'd expect it to be, and not the other fields. So let's just assert the handles and omit the other fields.

File: tests/auth.test.ts
Comment (shravanhariharan2) [reply tests/auth.test.ts:118]: This will also be useful for future test cases since we haven't fully written out our user registration test suite yet, so those tests would use also use this factory.

File: tests/user.handles.test.ts
Comment (shravanhariharan2) [inline tests/user.handles.test.ts:93]: Let's have these password changes be stored in a separate variable for improved readability, and let's add a comment on why these need to be included (i.e. mentioning that class-validator still validates optional fields for some reason and that it would error out if these fields aren't included).

File: tests/user.handles.test.ts
Comment (shravanhariharan2) [inline tests/user.handles.test.ts:102]: Excellent coverage here 👌🏾

File: tests/user.handles.test.ts
Comment (shravanhariharan2) [inline tests/user.handles.test.ts:107]: nit: since we use faker conventionally anyways, let's just use faker here instead to match the other tests:
```suggestion
    const MAX_LEN = 32;
    const user = UserFactory.fake({ handle: faker.datatype.hexadecimal(2 * MAX_LENGTH) });
```
Introducing randomness like this is useful to make it clear that the only thing we care about controlling is the length, and not the actual string.

File: tests/user.handles.test.ts
Comment (shravanhariharan2) [inline tests/user.handles.test.ts:114]: nit: same as above with moving the password changes to a separate variable (e.g. `const requiredValidatorFields = {...}`)

File: tests/user.handles.test.ts
Comment (shravanhariharan2) [inline tests/user.handles.test.ts:146]:
```suggestion
  test('Update handle error: too short and invalid characters', async () => {
```

**Review by shravanhariharan2 — COMMENTED** [review]
One main comment regarding the default handle values set by the migration. The rest are again mostly nitpicks around code quality and conventions

File: tests/auth.test.ts
Comment (farisashai) [reply tests/auth.test.ts:126]: lowkey copied the code from the other auth test that's doing the same thing with the hex string so idk what it means

File: migrations/0035-add-handle-column-to-users-table.ts
Comment (shravanhariharan2) [inline migrations/0035-add-handle-column-to-users-table.ts:26]: Is this indexing math correct? From [online](https://w3resource.com/PostgreSQL/substr-function.php) it seems like the third argument of `substr` is the number of characters to extract, which would mean that `firstName-lastName` would be 26 chars, the next dash would be 1 char, and then the final `substr` would be 7 chars, totalling 34 chars instead of 32. Have you verified that it's not 34 characters that's the limit here?

File: tests/auth.test.ts
Comment (shravanhariharan2) [reply tests/auth.test.ts:110]: Ok, let's just refactor that in a separate PR then. I think it's supposed to be the `accessCode` param.

**Review by shravanhariharan2 — APPROVED** [review]
Super great work 👍🏾 thanks for all the patience and for taking over the PR. One last just to make sure everything is correct, please address it before merging. Also, since this PR would cause a break without manual migration, let's make sure to coordinate that before merging.

**Review by shravanhariharan2 — CHANGES_REQUESTED** [review]
Revoking approval because we haven't dealt with this case yet:
> How do we deal with upper/lower case handle values? Should they be treated as identical if the stored database handle is "Faris-Ashai-123" and the API query param is "faris-ashai-123"?

Let's disallow uppercase for handles, similar to how Instagram does. We should also make the default handle created by the migration do only lower case as well.

File: migrations/0035-add-handle-column-to-users-table.ts
Comment (shravanhariharan2) [inline migrations/0035-add-handle-column-to-users-table.ts:26]: Is this setting the same length hash as the one that's generated on user signup? It looks like this one is of length 7, but the signup one is of length 6:
```ts
// Hexadecimals look like 0x1b9Dle so we have to truncate the fixed '0x'.
const hashValue = faker.datatype.hexaDecimal(6).slice(2);
```

Comment (farisashai) [reply migrations/0035-add-handle-column-to-users-table.ts:26]: Tested the migration again to confirm that the math is right - not sure why the substr arguments are so weird but it seems like its not inclusive for the start or end index so length is 25 (handle) + 5 hash + 2 dashes = 32

Screenshot 2023-03-29 at 6 31 26 PM

Comment (shravanhariharan2) [reply migrations/0035-add-handle-column-to-users-table.ts:26]: >not sure why the substr arguments are so weird but it seems like its not inclusive for the start or end index so length is 25 (handle) + 5 hash + 2 dashes = 32

It seems like the hashes here are of length 6 though? Looks like the handle including the first '-' is of length 25, the next dash is 1, and the hash is 6, resulting in 32 total. The reason why the start/end positioning might be weird is because I think Postgres does 1-based indexing (see example [here](https://www.postgresqltutorial.com/postgresql-string-functions/postgresql-substring/)), so starting at 0 may make it include an empty string at the start which includes an extra character. Otherwise, I'm unsure why specifying `26` makes the string length 25, even if the string is longer than 26 characters.

I hate to stall this PR even further, but can we quickly replace the 0-indexing with 1-indexing and see if this results in the same behavior? I.e. let's just replace the 0, 26 with 1, 25 and the 0, 7 with 1, 6 and verify the output. I think this will sharpen our understanding of Postgres and substring indexing, while also making the code less confusing. Appreciate the patience here

Comment (shravanhariharan2) [reply migrations/0035-add-handle-column-to-users-table.ts:26]: Will approve once this is verified.

**Review by shravanhariharan2 — APPROVED** [review]

**Review by steets250 — APPROVED** [review]
Dibs on `@steets250`

**Pull request merged**
